src/lib/http.coffee

```diff
@@ -21,12 +23,94 @@ module.exports = (dnschain) ->
+            # - https://github.com/openname/openname-specifications/blob/master/resolvers.md
+
+            opennameRoute = express.Router()
+
+            # Resolver specific API
+            opennameRoute.get /\/(?:resolver|dnschain)\/([^\/\.]+)(?:\.([a-z]+))?/, (req, res) =>
+                @log.debug gLineInfo("resolver API called"), {params: req.params}
+                [resource, format] = req.params
+                if resource == "fingerprint"
+                    if !format or format is 'json'
+                        res.json {fingerprint: @dnschain.encryptedserver.getFingerprint()}
+                    else
+                        @sendErr req, res, 400, "Unsupported format: #{format}"
+                else
+                    @sendErr req, res, 400, "Bad resource: #{resource}"
+
+            # Datastore API
+            opennameRoute.route(/// ^
+                \/(\w+)             # the datastore name
+                \/(\w+)             # the corresponding resource
+                (?:\/([^\/\.]+))?   # optional property (or action on resource)
+                (?:\/([^\/\.]+))?   # optional action on property
+                (?:\.([a-z]+))?     # optional response format
+                $ ///
+            ).get (req, res) =>
+                @log.debug gLineInfo("get v1"), {params: req.params}
+                [datastore, resource, propOrAction, _, _] = req.params
+                # this logic is here because the exact mechanism for retrieving
+                # the function to get resolver resources is not consistent across
+                # api versions, nor is the reason for failure to do so
+                resourceFnGetter = (resolver) ->
+                    [resolver.resources[resource], "Unsupported resource: #{resource}"]
+
+                @callback datastore, resourceFnGetter, ld.values(req.params).join(':')+":#{JSON.stringify(req.query)}", ld.values(req.params)[2..].concat(req.query), @postResolveCallback(req, res, propOrAction)
+
+            opennameRoute.use (req, res) =>
+                @sendErr req, res, 400, "Bad v1 request"
+
+            app.use "/v1", opennameRoute
+            app.get "*", (req, res) => # Old, deprecated API usage.
+                path = S(url.parse(req.originalUrl).pathname).chompLeft('/').s
+                options = url.parse(req.originalUrl, true).query
+                @log.debug gLineInfo('deprecated request'), {path:path, options:options, url:req.originalUrl}
+
+                [...,resolverName] =
+                    if S(header = req.headers.blockchain || req.headers.host).endsWith('.dns')
+                        S(header).chompRight('.dns').s.split('.')
+                    else
+                        ['none']
 
-            @server = http.createServer((req, res) =>
+                resourceFnGetter = (resolver) =>
+                    return [null, "Bad request: #{path}"] if not resolver.validRequest path
+                    [resolver.resolve, null]
```

Comment:
This `resourceFnGetter` is making things a bit weird. Something still feels off about it to me.

You can get rid of `resourceFnGetter` here and above (and all related code) by modifying `blockchain.coffee` (and all subclasses) to:

1. Get rid of `resolve:`
2. Move what was there into `resources.key`

That should make things clearer, cleaner, and get rid of some code duplication.